optaplanner-core/src/main/java/org/optaplanner/core/impl/processor/OptaPlannerAnnotationProcessor.java

```diff
@@ -0,0 +1,76 @@
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import javax.annotation.processing.AbstractProcessor;
+import javax.annotation.processing.ProcessingEnvironment;
+import javax.annotation.processing.RoundEnvironment;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.TypeElement;
+
+import org.optaplanner.core.api.domain.entity.PlanningEntity;
+import org.optaplanner.core.api.domain.variable.AnchorShadowVariable;
+import org.optaplanner.core.api.domain.variable.CustomShadowVariable;
+import org.optaplanner.core.api.domain.variable.InverseRelationShadowVariable;
+import org.optaplanner.core.api.domain.variable.PlanningVariable;
+
+public class OptaPlannerAnnotationProcessor extends AbstractProcessor {
+
+    @Override
+    public Set<String> getSupportedAnnotationTypes() {
+        return Collections.singleton("org.optaplanner.core.api.*");
+    }
+
+    private static final List<String> planningEntityMemberAnnotationList = Arrays.asList(
+            PlanningVariable.class.getSimpleName(),
+            InverseRelationShadowVariable.class.getSimpleName(),
+            AnchorShadowVariable.class.getSimpleName(),
+            CustomShadowVariable.class.getSimpleName());
+
+    private HashSet<String> planningEntityClassList;
+
+    @Override
+    public synchronized void init(ProcessingEnvironment processingEnv) {
+        super.init(processingEnv);
+        planningEntityClassList = new HashSet<>();
+    }
+
+    @Override
+    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
+        for (TypeElement annotation : annotations) {
+            if (planningEntityMemberAnnotationList
+                    .stream()
+                    .anyMatch(annotationName -> annotation.getSimpleName().contentEquals(annotationName))) {
+                processPlanningEntityMemberAnnotation(annotation, roundEnv);
+            }
+        }
+
+        return true;
+    }
+
+    private void processPlanningEntityMemberAnnotation(TypeElement annotation, RoundEnvironment roundEnv) {
+        Set<? extends Element> annotatedElements = roundEnv.getElementsAnnotatedWith(annotation);
+        annotatedElements.forEach(element -> {
+            switch (element.getKind()) {
+                case FIELD:
+                case METHOD:
+                    if (element.getEnclosingElement().getAnnotation(PlanningEntity.class) == null) {
+                        throw new IllegalStateException("Member annotation @" + annotation.getSimpleName() +
+                                " was placed in a class without @PlanningEntity ("
+                                + element.getEnclosingElement().getSimpleName()
+                                + "). Maybe add @PlanningEntity to the class (" + element.getEnclosingElement().getSimpleName()
+                                + ")?");
+                    }
```

Comment:
Does this work for example if the swing optaplanner-examples make this mistake?

Reply:
It will always work if `optaplanner-core` is on the classpath (as it contains the annotation processor entry). No additional configuration is required (at least for maven and gradle). If they are compiling with `javac`, some additional configuration might be required (since it doesn't generate any files, only verify proper annotation usage, everything will still work but the error won't be caught). It will also automatically work/be used for spring-boot. It doesn't seem to automatically work for quarkus though (at least in the deployment test), meaning quarkus probably does not automatically run annotation processors.